### ISSUE TITLE
BatchNorm3d error : CUDNN_STATUS_NOT_SUPPORTED

### ISSUE THREAD

**Issue by sdesrozis** [pytorch/pytorch#51776]
## 🐛 Bug

`BatchNorm3d` raises an error related to `cudnn`. 
```
RuntimeError: cuDNN error: CUDNN_STATUS_NOT_SUPPORTED. This error may appear if you passed in a non-contiguous input.
```

Thanks for the help or any advice to solve this !!

## To Reproduce

```
import torch
torch.backends.cuda.matmul.allow_tf32 = True
torch.backends.cudnn.benchmark = True
torch.backends.cudnn.deterministic = False
torch.backends.cudnn.allow_tf32 = True
data = torch.rand(70, 32, 100, 100, 100, dtype=torch.float16, device='cuda', requires_grad=False)
net = torch.nn.BatchNorm3d(32).to('cuda')
out = net(data)
print(out.shape)
```

## Expected behavior

The code above should print
```
torch.Size([70, 32, 100, 100, 100])
```

## Environment

```
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.9.2009 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.9 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-PCIE-16GB
GPU 1: Tesla V100-PCIE-16GB

Nvidia driver version: 440.33.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] pytorch-ignite==0.5.0
[pip3] torch==1.7.1
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.8.2
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] pytorch-ignite            0.5.0                    pypi_0    pypi
[conda] torch                     1.7.1                    pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.8.2                    pypi_0    pypi
```

**Comment by ngimel** [comment]
cudnn does not accept inputs with more than 2**31 elements.

**Comment by sdesrozis** [comment]
@ngimel Thank you for your answer. I didn't know about this limitation. I have seen a related answer you already did one year ago 

https://github.com/pytorch/pytorch/issues/22496#issuecomment-509433398

In the model I actually use, the batch norm is preceded by a con3d. Does PyTorch internally handle the exceedingly size by batch splitting as you mentioned ?

Thanks for your precious help!

**Comment by ngimel** [comment]
Yes, for convolutions it is easy to split a call to a convolution into multiple calls across the batch dimension, so that each input/output is less than 2**31 elements. Pytorch does this. For batchnorm however, since it has to normalize along the batch dimension, this splitting is non-trivial. 
@ptrblck Is cudnn going to support large inputs in the latest or future versions? If not, do you think it would make sense to send this case to native batch norm implementation?

**Comment by sdesrozis** [comment]
@ngimel Thank you again.

I hope that a solution can be made available either in `cudnn` or in `PyTorch`.

I was thinking that I could prototype with the idea of the actual implementation of `SyncBatchNorm`.

**Comment by ptrblck** [comment]
@sdesrozis Thanks for reporting this issue!

@ngimel I think cudnn should enable batchnorm layers to accept more than 2**31 elements (same for convs).
As it's not on their roadmap yet, I think falling back to the native implementation is the right approach at the moment.

If I'm not mistaken, a check for `input.nelement()` [here](https://github.com/pytorch/pytorch/blob/fe67438f327dc1b8b796159f200d7d5ae204740e/aten/src/ATen/native/Normalization.cpp#L417-L429) could enable the fallback?
If so, I can create the PR and run some tests.

**Comment by ngimel** [comment]
It could, but also look at the dtypes of parameters and running averages - I don't know if native can support half/float combination that cudnn supports.
